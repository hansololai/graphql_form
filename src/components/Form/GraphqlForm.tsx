import * as React from 'react';
import {
 Spin, notification, Form, Button,
} from 'antd';
import {
 FormInstance, Rule, FormItemProps, FormProps,
} from 'antd/lib/form';
import { updateInputQuery } from './queries';

// Generated Types
import { patchTypeQuery } from './__generated__/patchTypeQuery';
// import { fieldTypeQuery } from './__generated__/fieldTypeQuery';
import { useQueryWithError, createFormFields } from './utils';

import { validatorFunc } from './InputWrapper';
import 'antd/lib/form/style';
import 'antd/lib/button/style';
import 'antd/lib/notification/style';

const { useForm } = Form;

export interface FormFieldOptionProps {
  customValidators?: { [x: string]: validatorFunc };
  customRules?: { [x: string]: Rule[] };
  customDecorators?: { [x: string]: FormItemProps};
  customWidgets?: { [x: string]: React.SFC<{ value: any, onChange: (p: any) => void }> };
}

export interface GraphqlFormProps<FormData = any> extends FormFieldOptionProps, FormProps {
  modelName: string;
  instanceData?: FormData;
  instanceId?: number;
  onSubmit?: (form: FormInstance<FormData>) => void;
}
export const formItemLayout = {
  labelCol: {
    xs: { span: 24 },
    sm: { span: 8 },
  },
  wrapperCol: {
    xs: { span: 24 },
    sm: { span: 16 },
  },
};
const formLayout = {
  labelCol: { span: 8 },
  wrapperCol: { span: 16 },
};
const tailLayout = {
  wrappedCol: { offset: 0, span: 16 },
};

/**
 *
 */
export const GraphqlForm: React.SFC<GraphqlFormProps> = (props) => {
  const {
 modelName, instanceData, instanceId, onSubmit, ...options
} = props;
  const {
customValidators, customRules, customDecorators, customWidgets, ...formProps
} = options;
  // If has data, then it's update, otherwise it's a create form
  const typeName = instanceData ? `${modelName}Patch` : `${modelName}Input`;
  const [submitting, setSubmitting] = React.useState(false);
  const {
    data: inputData, loading: inputLoading,
  } = useQueryWithError<patchTypeQuery>(updateInputQuery, { variables: { name: typeName } });

  const inputFields = React.useMemo(() => inputData?.__type?.inputFields || [], [inputData]);
  const [form] = useForm<FormData>();
  React.useEffect(() => {
    if (instanceData) {
      form.setFieldsValue(instanceData);
    }
  }, [instanceData]);

  const allFields = createFormFields({
form,
inputFields,
customValidators,
    customRules,
customDecorators,
customWidgets,
});

  if (inputLoading) {
    return <Spin />;
  }

  return (
    <Form
      form={form}
      {...formLayout}
      onFinish={(values) => {
    setSubmitting(true);
    form.validateFields().then(() => {
      setSubmitting(false);
      if (onSubmit) {
        onSubmit(form);
      }
    }).catch((err) => {
      setSubmitting(false);
      notification.error({
        message: 'Validation Failed',
        description: err.message,
      });
    });
  }}
      initialValues={instanceData}
      {...formProps}
    >
      {allFields}
      {onSubmit
      && (
      <Form.Item {...tailLayout}>
        <Button type="primary" htmlType="submit" loading={submitting}>Submit</Button>
      </Form.Item>
)}
    </Form>
);
};
