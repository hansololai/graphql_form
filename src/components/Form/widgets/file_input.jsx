import React from 'react';
import { Upload, Button, Icon, Row } from 'antd';
import InputWrapper from './input_wrapper';


class UploadWithLink extends React.Component {
  render() {
    const { fileUrl, fileName, value } = this.props;
    const fileList = [];
    if (value instanceof File) {
      fileList.push(value);
    }
    return (
      <div>
        {fileUrl && fileName && <Row>
          <a href={fileUrl} >{fileName}</a>
        </Row>}
        <Upload name="logo" listType="picture" beforeUpload={() => false} {...this.props} fileList={fileList}>
          <Button>
            <Icon type="upload" /> Click to upload
          </Button>
        </Upload></div>);
  }
}
const FileInput = (props) => {
  const options = {
    getValueFromEvent: (e) => {
      console.log('Upload event:', e);
      if (Array.isArray(e)) {
        return e;
      }
      return e && e.file;
    },
  };
  const { fileUrl, fileName } = props;
  return (
    <InputWrapper {...props} options={options} >
      <UploadWithLink fileUrl={fileUrl} fileName={fileName} />
    </InputWrapper>);
};
export default FileInput;
