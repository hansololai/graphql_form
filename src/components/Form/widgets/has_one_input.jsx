import React from 'react';
import PropTypes from 'prop-types';
import { Select, notification } from 'antd';
import { Query, withApollo } from 'react-apollo';
import debounce from 'lodash.debounce';
import pluralize from 'pluralize';
import InputWrapper from './input_wrapper';
import { displayModelQL, MODELS } from '../../../graphql_queries';
import { isFunction } from '../../../utils';

const { Option } = Select;


class HasOneInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchInput: '',
      asyncFilter: {},
    };
    this.onSearch = this.onSearch.bind(this);
    this.setSearchInput = this.setSearchInput.bind(this);
    this.setFilter = this.setFilter.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.debounceSearch = debounce((text) => {
      this.setSearchInput(text);
    }, 200);
  }
  onFocus() {
    // Do the search only on focus
    this.setState({ focused: true });
    this.setSearchInput('');
  }
  onBlur() {
    this.setState({ focused: false });
  }

  onSearch(e) {
    this.debounceSearch(e);
  }
  setFilter() {
    const { asyncFilter, ...others } = this.props;
    if (isFunction(asyncFilter)) {
      asyncFilter(others).then((filter) => {
        this.setState({ asyncFilter: filter });
      });
    } else if (typeof (asyncFilter) === 'object') {
      this.setState({ asyncFilter });
    }
  }
  setSearchInput(text) {
    this.setFilter();
    this.setState({ searchInput: text });
  }
  render() {
    const { type, query, value, onChange } = this.props;
    const { searchInput, asyncFilter = {}, focused } = this.state;
    const notNull = type.kind === 'NOT_NULL';
    const info = type.kind === 'NON_NULL' ? type.ofType : type;
    const { name: typeName } = info;

    // Predefined fragment to query for models, it can either from props, or just the regular collection query
    const modelQuery = query || displayModelQL[typeName];
    if (!modelQuery) {
      // hmm no query, don't know how to render select if I don't know what field to display in options
      return <Select />;
    }

    const { searchfield } = MODELS.MODEL_HASH[typeName];
    // Set up variables, if there's a value, (which is the id) then it should be in the variables.
    // or if it matches the text in the search field
    const filter = { ...asyncFilter };
    if (focused) {
      // The user is doing a select
      const or = [];
      // if (Object.keys(asyncFilter).length > 0) {
      //   or.push(asyncFilter);
      // }
      if (searchInput) {
        or.push({
          [searchfield]: { includesInsensitive: searchInput },
        });
      }
      if (value) {
        or.push({
          id: { equalTo: Number(value) },
        });
      }
      filter.or = or;
    } else if (value) {
      // The dropdown is hiding, now only show the selected one
      filter.id = { equalTo: Number(value) };
    }
    return (<Query query={modelQuery} variables={{ filter, first: 50 }}>
      {({ loading, data, error }) => {
        if (loading) {
          return (<Select showSearch />);
        }
        if (error) {
          notification.error({
            message: 'error',
            description: error.message,
          });
          return null;
        }
        const pluralName = pluralize.plural(typeName);
        const { nodes } = data[`all${pluralName}`];

        const options = nodes.map(n => <Option value={n.id} key={n.id}>{n.name}</Option>);
        if (!notNull) {
          options.push(<Option value={null} key={'null'}>--</Option>);
        }
        return (
          <Select
            showSearch
            onSearch={this.onSearch}
            onChange={onChange}
            optionFilterProp="children"
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            value={value}
          >
            {options}
          </Select>);
      }}
    </Query>);
  }
}

HasOneInput.propTypes = {
  asyncFilter: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  type: PropTypes.object.isRequired,
  query: PropTypes.object,
};
HasOneInput.defaultProps = {
  query: null,
  value: undefined,
  asyncFilter: undefined,
};

const HasOneInputWrapper = (props) => {
  const { value: _, ...others } = props;
  return (<InputWrapper {...props}>
    <HasOneInput {...others} />
  </InputWrapper>)
  ;
};
HasOneInputWrapper.propTypes = {
  value: PropTypes.number,
};
HasOneInputWrapper.defaultProps = {
  value: undefined,
};
export default withApollo(HasOneInputWrapper);
