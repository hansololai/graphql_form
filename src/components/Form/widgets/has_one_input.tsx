import * as React from 'react';
import { Query } from 'react-apollo';
import { useDebounce } from 'use-debounce'
import { Select, Spin } from 'antd';
import { SelectFragmentProp } from '../ModelFragments';

export type NameFunction = (p: any) => string;
export interface HasOneInputProps<TData> extends SelectFragmentProp {
  value: any;
  onChange: (value: any) => void;
}

export const HasOneInput: <TData>(props: HasOneInputProps<TData>) => React.ReactElement<HasOneInputProps<TData>> = (props) => {
  const [searchInput, setSearchInput] = React.useState('');
  const [filterText] = useDebounce(searchInput, 500);
  const { selectQuery, value, onChange, nameField, valueField, filterField } = props;
  const filter = typeof filterField === "string" ? { [filterField]: { includesInsensitive: filterText } } : filterField(filterText);

  return <Query query={selectQuery} variables={{ first: 50, filter }}>
    {({ data, loading }) => {
      // Don't know what model it is, but it should have only 1 key
      let optionData: any[] = [];
      if (Object.values(data).length > 0) {
        const allModels: any = Object.values(data)[0]
        // if using simplify-infector, then it's probably don't have the "nodes", but let's assume this now
        if (allModels.nodes) {
          optionData = allModels.nodes;
        }
      }
      return <Select
        value={value}
        placeholder="type in to search"
        notFoundContent={loading ? <Spin size="small" /> : null}
        filterOption={false}
        onSearch={setSearchInput}
        onChange={onChange}
        style={{ width: '100%' }}
      >
        {optionData.map((opt) => {
          const name = (typeof nameField === 'string') ? opt[nameField] : nameField(opt);
          const optionValue = (typeof valueField === 'string') ? opt[valueField] : valueField(opt);
          return <Select.Option key={optionValue} value={optionValue}>{name}</Select.Option>
        })}
      </Select>;
    }}
  </Query>;
}