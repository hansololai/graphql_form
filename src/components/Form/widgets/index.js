import TextInput from './text_input';
import DateInput from './date_input';
import TimeInput from './time_input';
import NumberInput from './number_input';
import BooleanInput from './boolean_input';
import HasOneInput from './has_one_input';
import FileInput from './file_input';
import IntacctBillSelector from './intacct_bill_selector';
import TextSelectInput from './text_select_input';

export { TextInput, DateInput, TimeInput, FileInput, NumberInput, BooleanInput, HasOneInput, IntacctBillSelector, TextSelectInput };
