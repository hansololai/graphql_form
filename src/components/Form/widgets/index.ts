// import FileInput from './file_input';
// import IntacctBillSelector from './intacct_bill_selector';
export * from './scalar_input';
export * from './enum_input';
