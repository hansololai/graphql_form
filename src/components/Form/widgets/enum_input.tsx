import * as React from 'react';
import gql from 'graphql-tag';
import { Select, notification } from 'antd';
import { useQuery } from '@apollo/react-hooks';
import { enumTypeQuery, enumTypeQueryVariables } from './__generated__/enumTypeQuery';
import { WidgetProp } from '../GraphqlForm';

export interface EnumInputProps extends WidgetProp<any, string> {
  enumType: string;
}
export const enumQuery = gql`
  query enumTypeQuery($name: String!){
    __type(name: $name){
      name
      kind
      enumValues{
        name
      }
    }
  }
`;

/**
 * @description Input for Enum type, it is a select dropdown
 */
export const EnumInput: React.FC<EnumInputProps> = (props) => {
  const { value, onChange, enumType } = props;
  const {
    data, loading, error,
  } = useQuery<enumTypeQuery, enumTypeQueryVariables>(enumQuery, { variables: { name: enumType } });
  if (error) {
    notification.error({
      message: `Error Fetching options for ${enumType}`,
      description: error.message,
    });
  }
  const enumValues = React.useMemo(() => {
    const { __type: theType } = data || {};
    return theType?.enumValues || [];
  }, [data]);
  const options = enumValues.map((v) => (
    <Select.Option
      key={v.name}
      value={v.name}
    >
      {v.name}
    </Select.Option>
));
  return (
    <Select
      value={value}
      onChange={onChange}
      loading={loading}
    >
      {options}
    </Select>
);
};
