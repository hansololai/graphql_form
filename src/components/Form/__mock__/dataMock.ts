import { updateInputQuery, modelFieldsQuery } from '../queries';
import { enumQuery } from '../widgets/index';
import gql from 'graphql-tag'


export const sampleSelectQuery = gql`
query selectUser($first:Int, $filter:UserFilter){
  allUsers(first:$first, filter: $filter){
    nodes{
      id
      name
    }
  }
}
`;


export const mockTypeQueryData = {
  data: {
    __type: {
      __typename: "__Type",
      inputFields: [
        {
          "name": "id",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "Int",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "firstName",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "String",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "email",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "String",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "salary",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "Int",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "isAdmin",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "Boolean",
            "kind": "SCALAR",
            "ofType": null
          }
        },
      ]
    }
  }
}

export const mockCreateInputData = {
  data: {
    __type: {
      __typename: "__Type",
      inputFields: [
        {
          "name": "id",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "Int",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "firstName",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": null,
            "kind": "NON_NULL",
            "ofType": {
              "name": "String",
              "kind": "SCALAR",
              "__typename": "__Type"
            }
          }
        },
        {
          "name": "email",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": null,
            "kind": "NON_NULL",
            "ofType": {
              "name": "String",
              "kind": "SCALAR",
              "__typename": "__Type"
            }
          }
        },
        {
          "name": "salary",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",

            "name": "Int",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "isAdmin",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "Boolean",
            "kind": "SCALAR",
            "ofType": null
          }
        },
        {
          "name": "role",
          "__typename": "__InputValue",
          "defaultValue": null,
          "type": {
            "__typename": "__Type",
            "name": "UserRoleTypeEnum",
            "kind": "ENUM",
            "ofType": null
          }
        },
      ]
    }
  }
}

export const userRoleTypeEnumData = {
  "data": {
    "__type": {
      "name": "UserRoleTypeEnum",
      "__typename": "__Type",
      "kind": "ENUM",
      "enumValues": [
        {
          "name": "CEO",
          "__typename": "__EnumValue",
        },
        {
          "name": "CTO",
          "__typename": "__EnumValue",
        }
      ],
    }
  }
}
export const mockUserSelectData = {
  "data": {
    "allUsers": {
      "nodes": [
        { id: 1, name: "Kyle", __typename: "User" },
        { id: 2, name: "John", __typename: "User" },
        { id: 3, name: "George", __typename: "User" },
      ],
      __typename: "UserConnection"
    }
  }
}
export const mockUserSelectDataOnly = {
  "data": {
    "allUsers": {
      "nodes": [
        { id: 2, name: "John", __typename: "User" },
      ],
      __typename: "UserConnection"
    }
  }
}

export const mockUserModelFieldData = {
  "data": {
    "__type": {
      "__typename": "__Type",
      "fields": [
        {
          "__typename": "__Field",
          "name": "nodeId",
          "type": {
            "__typename": "__Type",
            "name": null,
            "kind": "NON_NULL",
            "ofType": {
              "__typename": "__Type",
              "name": "ID",
              "kind": "SCALAR"
            }
          }
        },
      ],
    },
  }
}


export const mockData = [
  {
    request: {
      query: modelFieldsQuery,
      variables: { name: "User" },
    },
    result: mockUserModelFieldData
  },
  {
    request: {
      query: sampleSelectQuery,
      variables: { first: 50, filter: { name: { includesInsensitive: "" } } }
    },
    result: mockUserSelectData,
  },
  {
    request: {
      query: sampleSelectQuery,
      variables: { first: 50, filter: { name: { includesInsensitive: "J" } } }
    },
    result: mockUserSelectDataOnly,
  },
  {
    request: {
      query: updateInputQuery,
      variables: { name: 'UserPatch' },
    },
    result: mockTypeQueryData
  },
  {
    request: {
      query: updateInputQuery,
      variables: { name: 'UserInput' },
    },
    result: mockCreateInputData
  },
  {
    request: {
      query: enumQuery,
      variables: { name: 'UserRoleTypeEnum' },
    },
    result: userRoleTypeEnumData
  }

]

