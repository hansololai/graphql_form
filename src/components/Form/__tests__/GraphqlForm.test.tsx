import * as React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { render, wait, fireEvent } from '@testing-library/react';
import { GraphqlForm } from '../GraphqlForm';
import { HasOneInput, TextSelectInput } from '../widgets';
import { mockData, sampleSelectQuery } from '../__mock__/dataMock'

const waitUntilNoSpin = (container: any) => wait(() => {
  const isLoading = container.querySelector('.anticon-spin') !== null;
  expect(isLoading).toBe(false);
});
describe('Graphql Form', () => {
  it('Basic Form', async () => {
    const onSubmit = jest.fn();
    const { container } = render(
      <MockedProvider mocks={mockData}>
        <div style={{ width: 400 }}>
          <GraphqlForm modelName="User" instanceData={{ id: 1, firstName: "test", email: "test@test.com" }} onSubmit={onSubmit} />
        </div>
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
    await waitUntilNoSpin(container);
    expect(container).toMatchSnapshot();
    // Try submit
    const formbutton = container.querySelector('button');
    expect(formbutton).not.toBeNull();
    fireEvent.click(formbutton);
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
  it('With new model data', async () => {
    const { container } = render(
      <MockedProvider mocks={mockData}>
        <div style={{ width: 400 }}>
          <GraphqlForm modelName="User" onSubmit={(form) => { }} />
        </div>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
    await waitUntilNoSpin(container);

    expect(container).toMatchSnapshot();
  });
  it('With Custom Widget and Rule', async () => {
    const customWidgets = {
      firstName: ({ value, onChange }) => <TextSelectInput value={value} onChange={onChange} inputOptions={[
        { name: 'John(Manager)', value: 'John' },
        { name: 'Robin(CEO)', value: 'Robin' },
        { name: 'Evan(Intern)', value: 'Evan' },
      ]} />
    };
    const customRule = {
      pattern: /@/
    };
    const customValidator = (rule, value, cb, form) => {
      // call cb() means no error
      // call cb("error message") means there is error
      if (Number(value) >= 10000 && Number(value) <= 50000) {
        cb();
      }
      cb("not within 10000 - 50000");
    };

    const { container } = render(
      <MockedProvider mocks={mockData}>
        <div style={{ width: 400 }}>
          <GraphqlForm modelName="User" onSubmit={(form) => { }}
            customWidgets={customWidgets}
            customRules={{
              email: [customRule]
            }}
            customValidators={{
              salary: customValidator
            }}
          />
        </div>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
    await waitUntilNoSpin(container);
    expect(container).toMatchSnapshot();
  });
  it('With Enum type', async () => {
    const { container } = render(
      <MockedProvider mocks={mockData}>
        <div style={{ width: 400 }}>
          <GraphqlForm modelName="User" onSubmit={(form) => { }} />
        </div>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();

    await waitUntilNoSpin(container);
    expect(container).toMatchSnapshot();
  });
})
it('SelectWidget of user', async () => {
  const { container } = render(
    <MockedProvider mocks={mockData}>
      <HasOneInput selectQuery={sampleSelectQuery} nameField="name" valueField="id" filterField="name" value={null} onChange={(e) => { }} />

    </MockedProvider>
  )
  expect(container).toMatchSnapshot();

  await waitUntilNoSpin(container);
  expect(container).toMatchSnapshot();
})