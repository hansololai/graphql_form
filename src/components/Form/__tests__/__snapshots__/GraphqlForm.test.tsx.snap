// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Graphql Form 1`] = `
<div>
  <div
    style="width: 400px;"
  >
    <div
      class="ant-spin ant-spin-spinning"
    >
      <span
        class="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
      </span>
    </div>
  </div>
</div>
`;
